src/com/ms/silverking/net/async/AsyncBase.java

```diff
@@ -263,160 +258,180 @@ public void doWork(Connection connection) {
+          // If the reconnect fails, then it will get marked.
+          //if (suspectAddressListener != null) {
+          //  suspectAddressListener.addSuspect(connection.getRemoteSocketAddress());
+          //} else {
+          //  Log.warning("suspectAddressListener is null");
+          //}
         }
-        
-        @Override
-        public void doWork(Connection connection) {
-            boolean cleanWrite;
-
-            if (AsyncGlobals.debug) {
-                Log.fine("AsyncBase.Writer.doWork()");
-            }
-            cleanWrite = false;
-            try {
-                connection.writeAllPending();
-                cleanWrite = true;
-            } catch (IOException ioe) {
-                if (connection.isConnected()) {
-                    Log.logErrorWarning(ioe);
-                    // Commented out for new approach.
-                    // New approach is to simply do the disconnect as before, but not mark as suspect.
-                    // If the reconnect fails, then it will get marked.
-                    //if (suspectAddressListener != null) {
-                    //  suspectAddressListener.addSuspect(connection.getRemoteSocketAddress());
-                    //} else {
-                    //  Log.warning("suspectAddressListener is null");
-                    //}
-                }
-            } catch (Exception e) {
-                if (connection.isConnected()) {
-                    Log.logErrorWarning(e, "Unhandled exception()");
-                }
-            } finally {
-                if (!cleanWrite) {
-                    disconnect(connection, "!cleanWrite");
-                }
-            }
+      } catch (Exception e) {
+        if (connection.isConnected()) {
+          Log.logErrorWarning(e, "Unhandled exception()");
         }
-    }
-                
-    ////////////////////////////////////////////////////////////////////////////////////
-    public T newOutgoingConnection(InetSocketAddress dest, ConnectionListener listener) throws IOException, ConnectionAbsorbException {
-        SocketChannel    channel;
-        
-        channel = SocketChannel.open();
-        if (dest.isUnresolved()) {
-            Log.warning("Unresolved InetSocketAddress: "+ dest);
-            throw new ConnectException("Unresolved InetSocketAddress"+ dest.toString());
+      } finally {
+        if (!cleanWrite) {
+          disconnect(connection, "!cleanWrite");
         }
-        LWTThreadUtil.setBlocked();
+      }
+    }
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////////
+  public T newOutgoingConnection(InetSocketAddress dest, ConnectionListener listener)
+      throws IOException, ConnectionAbsorbException {
+    SocketChannel channel = null;
+    boolean connectionSuccess = false;
+
+    try {
+      channel = SocketChannel.open();
+      LWTThreadUtil.setBlocked();
+      channel.socket().connect(dest, defSocketConnectTimeout);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *socket*. The resource is referenced in statements at the following lines: 319, 325, 326, 333, 335, 339. The resource closure statement is at line: 337. There are other execution paths that do not contain closure statements, for example, when Socket.connect throws an exception, and an else-branch is taken at line 335. Either a) close the object returned by *socket()* in a try-finally block or b) close the resource by declaring the object returned by *socket()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).